proposals/Layer-1.md

```diff
@@ -0,0 +1,390 @@
+block is a sequence of instructions following the `if_except` instruction. The
+else block is optional, and if it appears, it begins with the `else`
+instruction.  The scope of the if_except block is from the `if_except`
+instruction to the corresponding `end` instruction.
+
+That is, the forms of an if_except block is:
+
+```
+if_except block_type except_index
+  Instruction*
+end
+
+if_except block_type except_index
+  Instruction*
+else
+  Instructions*
+end
+```
+
+The conditional query of an exception succeeds when the exception on the top of
+the stack is an instance of the corresponding tagged exception type (defined by
+`except_index`).
+
+If the query succeeds, the data values (associated with the type signature of
+the exception class) are extracted and pushed onto the stack, and control
+transfers to the instructions in the then block.
+
+If the query fails, it either enters the else block, or transfer control to the
+end of the if_except block if there is no else block.
+
+### Debugging
+
+Earlier discussion implied that when an exception is thrown, the runtime will
+pop the operand stack across function calls until a corresponding, enclosing try
+block is found. The implementation may actually not do this. Rather, it may
+first search up the call stack to see if there is an enclosing try. If none are
+found, it could terminate the thread at the point of the throw. This would
+allow better debugging capability, since the corresponding call stack is still
+there to query.
+
+## Changes to the text format.
+
+This section describes change in the
+[instruction syntax document](https://github.com/WebAssembly/spec/blob/master/document/core/instructions.rst).
+
+### New instructions
+
+The following rules are added to *instructions*:
+
+```
+  try resulttype instructions* catch instructions* end |
+  except except_index |
+  throw |
+  if_except resulttype except_index then Instructions* end |
+  if_except resulttype except_index then Instructions* else Instructions* end
+```
+
+Like the `block`, `loop`, and `if` instructions, the `try` and `if_except`
+instructions are *structured* control flow instructions, and can be
+labeled. This allows branch instructions to exit try and `if_except` blocks.
+
+The `except_index` of the `except` and `if_except` instructions defines the
+exception type to create/extract form. See [exception index
+space](#exception-index-space) for further clarification of exception tags.
+
+## Changes to Modules document.
+
+This section describes change in the
+[Modules document](https://github.com/WebAssembly/design/blob/master/Modules.md).
+
+### Exception index space
+
+The _exception index space_ indexes all imported and internally-defined
+exceptions, assigning monotonically-increasing indices based on the order
+defined in the import and exception sections. Thus, the index space starts at
+zero with imported exceptions followed by internally-defined exceptions in
+the [exception section](#exception-section).
```

Comment:
1. Are 'exception index' and 'exception tag' the same thing?
2. If the indices of imported exceptions start from 0, how does a wasm compiler know which tag to use to denote, let's say, C++ exception? Is it the linker's job to reconcile all the indices and fix all indices within a wasm program?

Reply:
> 1. Are 'exception index' and 'exception tag' the same thing?

No, but they're similar. The exception tag is an opaque identifier for the type of the exception. The exception index points to an entry in the exception table, which has its own tag.

> 2. If the indices of imported exceptions start from 0, how does a wasm compiler know which tag to use to denote, let's say, C++ exception? Is it the linker's job to reconcile all the indices and fix all indices within a wasm program?

By wasm compiler, do you mean something like the C++ to Wasm compiler (as opposed to the Wasm to Native Code VM)? If so, then the linker seems like a reasonable place to reconcile all the exception indices. Ultimately, the notion of "the C++ exception type" does not really exist in the Wasm spec, but is a convention used by the tools.

Reply:
> The exception tag is an opaque identifier for the type of the exception. The exception index points to an entry in the exception table, which has its own tag.

Then an `except` instruction take an exception index as an immediate, right? In the Binary encoding section, it says `tag`, so it was confusing.